Firestore/core/src/firebase/firestore/core/transaction.mm

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/core/transaction.h"
+
+#include <algorithm>
+#include <unordered_set>
+#include <utility>
+
+#import "Firestore/Source/Model/FSTDocument.h"
+#import "Firestore/Source/Model/FSTMutation.h"
+
+#include "Firestore/core/include/firebase/firestore/firestore_errors.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+using firebase::firestore::FirestoreErrorCode;
+using firebase::firestore::core::ParsedSetData;
+using firebase::firestore::core::ParsedUpdateData;
+using firebase::firestore::model::DocumentKey;
+using firebase::firestore::model::DocumentKeyHash;
+using firebase::firestore::model::Precondition;
+using firebase::firestore::model::SnapshotVersion;
+using firebase::firestore::remote::Datastore;
+using firebase::firestore::util::Status;
+using firebase::firestore::util::StatusOr;
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+Transaction::Transaction(Datastore* datastore)
+    : datastore_{NOT_NULL(datastore)} {
+}
+
+Status Transaction::RecordVersion(FSTMaybeDocument* doc) {
+  SnapshotVersion doc_version;
+
+  if ([doc isKindOfClass:[FSTDocument class]]) {
+    doc_version = doc.version;
+  } else if ([doc isKindOfClass:[FSTDeletedDocument class]]) {
+    // For deleted docs, we must record an explicit no version to build the
+    // right precondition when writing.
+    doc_version = SnapshotVersion::None();
+  } else {
+    HARD_FAIL("Unexpected document type in transaction: %s",
+              NSStringFromClass([doc class]));
+  }
+
+  absl::optional<SnapshotVersion> existing_version = GetVersion(doc.key);
+  if (existing_version.has_value()) {
+    if (doc_version != existing_version.value()) {
+      return Status{// This transaction will fail no matter what.
```

Comment:
As discussed, this changes the behavior -- previously, iOS would unconditionally forbid reading the same document more than once. The new behavior is same as other platforms.